proposal/first-class-calc.md

```diff
@@ -0,0 +1,541 @@
+
+[`min()` and `max()` functions]: ../accepted/min-max.md
+
+Here are some alternatives we considered:
+
+1. Re-parsing a calc that contains interpolation once the interpolation has been
+   resolved, and using the result as in the calc object rather than an unquoted
+   string. For example, `calc(#{"1px + 2px"})` would return `3px` rather than
+   `calc(1px + 2px)`. However, doing another parse at evaluation-time would add
+   substantial complexity and some amount of runtime overhead. The
+   return-on-investment would also be inherently limited, since we're planning
+   on gradually transitioning users away from interpolation in `calc()` anyway.
+
+2. Treating interpolation as a `CalcValue` that participates in the normal
+   parsing flow of a `CalcArgument`. This is a simpler and more efficient method
+   since it doesn't require parser lookahead, and it supports common cases like
+   `calc(#{$var} + 10%)` well. However, it doesn't support cases like `calc(1px
+   #{$op} 10%)` which are currently supported. This backwards-incompatibility is
+   likely to cause real user pain for a feature as widely-used as `calc()`.
+
+#### Vendor Prefixed `calc()`
+
+Although `calc()` is now widely supported in all modern browsers, older versions
+of Firefox, Chrome, and Safari supported it only with a vendor prefix. Sass in
+turn supported those browsers by handling `calc()`'s special function parsing
+with arbitrary vendor prefixes as well. However, time has passed, those browser
+versions have essentially no usage any more, and we don't anticipate anyone is
+looking to write new stylesheets that target them.
+
+As such, this proposal only adds first-class calc support for the `calc()`
+function without any prefixes. For backwards-compatibility, vendor-prefixed
+`calc()` expressions will continue to be parsed as opaque special functions the
+way they always have, but they will not be interoperable with any of the new
+calc features this proposal adds.
+
+## Syntax
+
+### `SpecialFunctionExpression`
+
+This proposal replaces the definition of [`SpecialFunctionName`] with the
+following:
+
+[`SpecialFunctionName`]: ../spec/syntax.md#specialfunctionexpression
+
+<x><pre>
+**SpecialFunctionName**¹      ::= VendorPrefix? ('element(' | 'expression(')
+&#32;                           | VendorPrefix `calc('
+</pre></x>
+
+1: The string `calc(` is matched case-insensitively.
+
+### `CalcExpression`
+
+This proposal defines a new production `CalcExpression`. This expression is
+parsed in a SassScript context when an expression is expected and the input
+stream starts with an identifier with value `calc` (ignoring case) followed
+immediately by `(`.
+
+The grammar for this production is:
+
+<x><pre>
+**CalcExpression** ::= `calc(`¹ CalcArgument ')'
+**ClampExpression** ::= `clamp(`¹ CalcArgument ',' CalcArgument ',' CalcArgument ')'
+**CalcArgument**²  ::= InterpolatedDeclarationValue | CalcSum
+**CalcSum**     ::= CalcProduct (('+' | '-')³ CalcProduct)?
```

Comment:
shouldn't the right hand side actually be `CalcSum` to support `1px + 1% + 1px` ? the current syntax would require braces to achieve that as `1px + (1% + 1px). (making the `CalcProduct` a single `CalcValue` containing a `CalcArgument` enclosed in braces).

Reply:
You're right that this is wrong, but I actually intended to make the second clause repeated rather than using recursion to repeat them (that's the structure that's assumed in the Semantics section below).

Reply:
then the simplification algorithm needs to be updated, as it was describing only operations with 2 terms.

Reply:
Actually, `CalculationOperation` only has `left` and `right` so the necessary change is explaining how this repeated clause gets turned into the struct.

Reply:
I think it works as written, the `CalcSum` and `CalcProduct` semantics sections assemble the lists into a sequence of `CalcOperations` with two operands each.